### PR TITLE
polish ukernel test cpu features

### PR DESCRIPTION
**Pull request by bjacob**
* Support comma-separated lists of features (existing TODO)
* Add `util_test`, test the mapping of cpu feature strings to bits --- had gotten nontrivial enough, and a bug there could silently affect test coverage.
* Use empty string instead of NULL for no CPU features -- we had 2 null-values here, should have only one. That part was #13229.
* Make `base/internal/cpu.h` no longer include `iree/schemas/cpu_data.h` - it doesn't need it anymore and it's best to have users not rely on it to include that for them.

### CODE REVIEW

File: runtime/src/iree/builtins/ukernel/tools/util_test.c [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  // Comma-separated lists of arm features.
+  expected[0] = IREE_CPU_DATA0_ARM_64_DOTPROD | IREE_CPU_DATA0_ARM_64_I8MM;
+  iree_uk_test_make_cpu_data_for_features_case(test, "dotprod,i8mm", expected);
+  // Named arm64 feature sets: none at the moment.
+
+#endif  // defined(IREE_UK_ARCH_X86_64)
+}
+
+int main(int argc, char** argv) {
+  iree_uk_test("make_cpu_data_for_features",
```
Comment (benvanik) [inline runtime/src/iree/builtins/ukernel/tools/util_test.c:70]: you need to return something here

**Review by benvanik — CHANGES_REQUESTED** [review]

File: runtime/src/iree/builtins/ukernel/tools/util_test.c (hunk 1 above)
Comment (bjacob) [reply runtime/src/iree/builtins/ukernel/tools/util_test.c:70]: Oh wow, TIL - and I have to make that fix in several files now. I thought that hitting the end of `main` implicitly returned `0`. Apparently that was the case before C99 but not anymore as of C99. https://en.cppreference.com/w/c/language/main_function

Comment (bjacob) [reply runtime/src/iree/builtins/ukernel/tools/util_test.c:70]: Ah no! It's actually still specified as `return 0` as long as the return type is `int`.  The only thing that changed in C99 was in the case of `main` function returning another type than `int`.

Comment (bjacob) [reply runtime/src/iree/builtins/ukernel/tools/util_test.c:70]: https://en.cppreference.com/w/c/language/main_function

* until C99:
   * If the main function executes a return that specifies no value or, which is the same, reaches the terminating } without executing a return, the termination status returned to the host environment is undefined.
* since C99:
  * If the return type of the main function is not compatible with int (e.g. void main(void)), the value returned to the host environment is unspecified.
  * If the return type is compatible with int and control reaches the terminating }, the value returned to the environment is the same as if executing return 0;.

Comment (benvanik) [reply runtime/src/iree/builtins/ukernel/tools/util_test.c:70]: What works and what is readable are not always aligned :) I'd prefer straightforward readable code vs. corner cases in the spec as understanding what a program returns should be easy to spot. I'd expect a test to return 0 or non-zero (EXIT_SUCCESS or EXIT_FAILURE) based on its results and the way this is currently setup is non-obvious. I see you're just calling iree_abort on failure, which is not great - on windows that'll pop up a modal dialog in debug builds and on others it can result in coredumps and other unexpected weirdness that makes automation more difficult. It'd be good to fix that in the future but for now adding `return EXIT_SUCCESS;  // iree_abort is used on failure` or something would help readers understand the current behavior.

**Review by benvanik — APPROVED** [review]

File: runtime/src/iree/builtins/ukernel/tools/util_test.c (hunk 1 above)
Comment (bjacob) [reply runtime/src/iree/builtins/ukernel/tools/util_test.c:70]: Added `return EXIT_SUCCESS;`. Interesting points. Here's my take.
- **Testability.** As the `return 0` is implied per spec and the compiler, correspondingly, does not warn on the missing `return`, if we do care about consistently explicitly `return 0`, we are on our own, the compiler won't help us. Which the pessimistic in me takes to entail that we will *not* have `return 0` consistently over time. We'll forget about it, or some contributor will, and we'll forget requesting it in review, etc.
- **Non-fatal failures** have costs:
  * *Implementation:* Either (1) the testing code becomes stateful, or (2) one passes some data structure from the inner testcases code, where non-fatal failures occur, all the way back to the `main` function to decide its return value. State is magic so (1) is scary, and passing around data structures is the definition of bureaucracy so (2) is scary too. I'm trying to do the least amount of scary things for my immediate needs.
  * *Usage:* When a test failure occurs, it's easiest to debug if it's fatal, so one simply runs test test in a debugger without even having to set a breakpoint.
- **Aborting** is what happens anyway on assertion failure, which is always common in tests; even more so here as ukernels validation is assertions. Since it will be a common test failure more anyway, I think it makes sense to keep all test failures be aborts. If aborts have annoying behavior on some platform then we have a problem as they will occur anyway.

Comment (benvanik) [reply runtime/src/iree/builtins/ukernel/tools/util_test.c:70]: Since you engaged the philosophical discussion... ;)

Mostly musing - you can keep what you have, but I do think this is an interesting topic with the focus mostly about avoiding surprising/non-standard unit test behavior - see https://en.wikipedia.org/wiki/Principle_of_least_astonishment. We expect other people to use this stuff even if they aren't familiar with it (overwhelmed integrators, people bringing up new platforms, etc) and a good way to help with that is by being unsurprising. Technical aspects can still win over that and lead to intentionally surprising behavior but they have a much higher bar (like that we need to run these tests on all platforms that may execute these kernels and that's why we don't just use gtest, which is surprising but required). There are fine lines on all this stuff but those are more interesting to walk when it's something that matters to the project - test infra should be boring and the easiest way to be boring is to do what everyone else does :)

So since we use gtest in the rest of the project we can say that unless there's a good reason the other kinds of unit tests we have should behave roughly the same in order to be least surprising to people familiar with the project. gtest's default main returns an exit code and the custom main `RUN_ALL_TESTS` macro returns an exit code to return from main - it doesn't abort on first failure or trigger a coredump on any ASSERT or EXPECT failure. This happens to also be similar with cunit, catch, and all other popular options (I tried running down the wikipedia's c unit test framework list and finding any that didn't but got bored lol), and test runners in java/js/etc. By differing from this behavior the large majority of people who have used any command line test framework would be surprised and need time to understand what was happening. It may not be a big surprise but small surprises add up when the principle isn't applied consistently. Most people would rather fix the issue and move on then weigh the pros/cons of divergent behavior :)

Why this behavior is surprising is that unit tests by nature should not abort on first failure unless there's an actual crash: in the steady state unit test runners are about getting the results of all unit tests and then presenting those as there is often information to be learned from failure patterns. Knowing that all f32 mmt4d tests fail but all i8 ones pass is useful in triaging the issue but not something you'd see if it aborted on the first f32 test and never got to the rest. Most users and CIs (which is where most users will interact with these tests) are best served with keep-going behavior instead of stop-on-first-error - and it's why all test frameworks keep going by default; nobody wants stop-on-first on ctest runs and CIs:
simpsons-sideshow-bob 1

Test frameworks that do support scoped fatal assertions (like gtest ASSERT) do so by setting a flag, returning early from the test, and then continuing on with the other unit tests. We're not building a complex test framework doing concurrent or repeated test execution so setting a global in `iree_uk_test_fail` and returning that from an `iree_uk_test_return_code()` would be safe and easy for this scale of thing - effectively just `return any_test_failures ? EXIT_FAILURE : EXIT_SUCCESS;`. It's not scary at this scale - it's the first stuff taught in CS. I _wouldn't_ do that in the ukernel library but in a tiny test runner? #yolo in a rwdata `bool`

But aborting on first failure may be useful for you while debugging with your workflow. There's a few ways to do that - easiest being `break iree_uk_test_fail` if you're about debugging. For programmatic control gtest uses the flag `--gtest_break_on_failure` but that'd need flags parsing. A libc only dependency we could take is `if (getenv("BREAK_ON_FAILURE")) iree_abort();` and you can print that env var out if any test fails before exiting so it's easy for users to discover (only finding it after their first failure). This lets users opt-in when they are doing interactive iterative development and otherwise be standard and good for CI/drive-bys.

Mostly just want to convey that it's not the technical aspects but the divergent behavior at issue: the standard behavior is that unit test runners run as many tests as possible in a single invocation and indicate which fail, return with a non-zero exit code if any test fails, and provide a way to opt-in to stop-on-first-error if someone wants it. Anything else is surprising and we don't want surprises in our test infra :)

Comment (bjacob) [reply runtime/src/iree/builtins/ukernel/tools/util_test.c:70]: OK, thanks for the detailed write-up! I can do that -- and that's just the right level of complexity for my fried brain tonight.
